### PR TITLE
UCT/IB/MLX5: DEVX KSM

### PR DESCRIPTION
**Pull request by Artemy-Mellanox**

### CODE REVIEW

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/8428/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/8443/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/8607/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/8608/ for details (Mellanox internal link).

**Comment by Artemy-Mellanox** [comment]
@yosefe

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/8609/ for details (Mellanox internal link).

File: src/uct/ib/mlx5/ib_mlx5_dv.c [hunk 1]
```diff
@@ -106,7 +134,10 @@ static ucs_status_t uct_ib_mlx5dv_device_init(uct_ib_device_t *dev)
                (errno != EPROTONOSUPPORT) &&
                (errno != EOPNOTSUPP)) {
         ucs_error("MLX5_CMD_OP_QUERY_HCA_CAP failed: %m");
-        return UCS_ERR_IO_ERROR;
+        status = UCS_ERR_IO_ERROR;
+        goto err;
+    } else {
+        return UCS_ERR_UNSUPPORTED;
```
Comment (brminich) [inline src/uct/ib/mlx5/ib_mlx5_dv.c]: goto err?

File: src/uct/ib/base/ib_device.h [hunk 2]
```diff
@@ -61,7 +61,8 @@ enum {
     UCT_IB_DEVICE_FLAG_DC_V2    = UCS_BIT(7),   /* Device supports DC ver 2 */
     UCT_IB_DEVICE_FLAG_AV       = UCS_BIT(8),   /* Device supports compact AV */
     UCT_IB_DEVICE_FLAG_DC       = UCT_IB_DEVICE_FLAG_DC_V1 |
-                                  UCT_IB_DEVICE_FLAG_DC_V2 /* Device supports DC */
+                                  UCT_IB_DEVICE_FLAG_DC_V2, /* Device supports DC */
+    UCT_IB_DEVICE_FLAG_KSM      = UCS_BIT(9),   /* Device supports KSM */
```
Comment (brminich) [inline src/uct/ib/base/ib_device.h]: extra comma

File: src/uct/ib/base/ib_md.c [hunk 3]
```diff
@@ -403,13 +403,15 @@ static ucs_status_t uct_ib_md_reg_mr(uct_ib_md_t *md, void *address,
...
-                                       off_t offset, struct ibv_mr **umr_p)
+static ucs_status_t uct_ib_verbs_md_post_umr(uct_ib_md_t *md,
+                                             uct_ib_mem_t *memh,
+                                             off_t offset)
 {
 #if HAVE_EXP_UMR
     struct ibv_exp_mem_region *mem_reg = NULL;
     struct ibv_exp_send_wr wr, *bad_wr;
     struct ibv_exp_create_mr_in mrin;
+    struct ibv_mr *mr = memh->mr;
```
Comment (brminich) [inline src/uct/ib/base/ib_md.c]: initialized vars go first

File: src/uct/ib/mlx5/ib_mlx5_dv.c [hunk 4]
```diff
@@ -24,6 +33,110 @@ ucs_status_t uct_ib_mlx5dv_init_obj(uct_ib_mlx5dv_t *obj, uint64_t type)
...
 #if HAVE_DECL_MLX5DV_CONTEXT_FLAGS_DEVX
+static ucs_status_t uct_ib_mlx5dv_create_ksm(uct_ib_md_t *md,
+                                             uct_ib_mem_t *ib_memh,
+                                             off_t offset)
+{
+    uct_ib_mlx5_mem_t *memh = ucs_derived_of(ib_memh, uct_ib_mlx5_mem_t);
+    uint32_t out[UCT_IB_MLX5DV_ST_SZ_DW(create_mkey_out)] = {0};
+    struct ibv_mr *mr = memh->super.mr;
+    size_t reg_length, length, inlen;
+    ucs_status_t status = UCS_OK;
```
Comment (brminich) [inline src/uct/ib/mlx5/ib_mlx5_dv.c]: initialized vars go first

File: src/uct/ib/mlx5/ib_mlx5_dv.c [hunk 5]
```diff
@@ -24,6 +33,110 @@ ucs_status_t uct_ib_mlx5dv_init_obj(uct_ib_mlx5dv_t *obj, uint64_t type)
...
+    int list_size, i;
+    void *mkc, *klm;
+    uint32_t *in;
+    intptr_t addr;
+
+    if (!(md->dev.flags & UCT_IB_DEVICE_FLAG_KSM)) {
+        return UCS_ERR_UNSUPPORTED;
+    }
+
+    reg_length = UCT_IB_MD_MAX_MR_SIZE;
```
Comment (brminich) [inline src/uct/ib/mlx5/ib_mlx5_dv.c]: minor: can align by `=`

File: src/uct/ib/mlx5/ib_mlx5_ifc.h [hunk 6]
```diff
@@ -58,7 +58,8 @@
     be64toh(*((__be64 *)(_p) + __uct_64_off(_typ, _fld)))
 
 enum {
-    UCT_IB_MLX5_CMD_OP_QUERY_HCA_CAP        = 0x100
+    UCT_IB_MLX5_CMD_OP_QUERY_HCA_CAP        = 0x100,
+    UCT_IB_MLX5_CMD_OP_CREATE_MKEY          = 0x200,
```
Comment (brminich) [inline src/uct/ib/mlx5/ib_mlx5_ifc.h]: extra comma

File: src/uct/ib/mlx5/ib_mlx5_ifc.h [hunk 7]
```diff
@@ -434,4 +435,110 @@ struct uct_ib_mlx5_query_hca_cap_in_bits {
     uint8_t    reserved_at_40[0x40];
 };
 
+enum {
+    UCT_IB_MLX5_MKC_ACCESS_MODE_PA    = 0x0,
+    UCT_IB_MLX5_MKC_ACCESS_MODE_MTT   = 0x1,
+    UCT_IB_MLX5_MKC_ACCESS_MODE_KLMS  = 0x2,
+    UCT_IB_MLX5_MKC_ACCESS_MODE_KSM   = 0x3,
+    UCT_IB_MLX5_MKC_ACCESS_MODE_MEMIC = 0x5,
```
Comment (brminich) [inline src/uct/ib/mlx5/ib_mlx5_ifc.h]: extra comma

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/8610/ for details (Mellanox internal link).

**Comment by Artemy-Mellanox** [comment]
bot:mlx:retest

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/8629/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/8635/ for details (Mellanox internal link).

**Comment by Artemy-Mellanox** [comment]
bot:mlx:retest

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/8639/ for details (Mellanox internal link).

File: src/uct/ib/base/ib_md.c [hunk 8]
```diff
@@ -583,13 +593,13 @@ static ucs_status_t uct_ib_dereg_mr(struct ibv_mr *mr)
...
 
-static ucs_status_t uct_ib_memh_dereg(uct_ib_mem_t *memh)
+static ucs_status_t uct_ib_memh_dereg(uct_ib_md_t *md, uct_ib_mem_t *memh)
 {
     ucs_status_t s1, s2;
 
     s1 = s2 = UCS_OK;
     if (memh->flags & UCT_IB_MEM_FLAG_ATOMIC_MR) {
-        s2 = uct_ib_dereg_mr(memh->atomic_mr);
+        s2 = md->ops->dereg_atomic_key(md, memh);
```
Comment (brminich) [inline src/uct/ib/base/ib_md.c]: why don't we define a wrapper, like `uct_ib_reg_atomic_key`? Would be consistent with other funcs

File: src/uct/ib/mlx5/ib_mlx5_dv.c (hunk 4 above)
Comment (brminich) [reply src/uct/ib/mlx5/ib_mlx5_dv.c]: it is still relevant, vars above are not initialized

File: src/uct/ib/base/ib_md.c (hunk 8 above)
Comment (brminich) [reply src/uct/ib/base/ib_md.c]: ok, ic

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/8642/ for details (Mellanox internal link).

**Comment by Artemy-Mellanox** [comment]
@yosefe

File: src/uct/ib/mlx5/ib_mlx5_dv.c [hunk 9]
```diff
@@ -24,6 +33,110 @@ ucs_status_t uct_ib_mlx5dv_init_obj(uct_ib_mlx5dv_t *obj, uint64_t type)
...
+    UCT_IB_MLX5DV_SET(mkc, mkc, translations_octword_size, list_size);
+    UCT_IB_MLX5DV_SET(mkc, mkc, log_entity_size, ucs_ilog2(reg_length));
+    UCT_IB_MLX5DV_SET(mkc, mkc, qpn, 0xffffff);
+    UCT_IB_MLX5DV_SET(mkc, mkc, mkey_7_0, offset & 0xff);
+    UCT_IB_MLX5DV_SET64(mkc, mkc, start_addr, addr + offset);
+    UCT_IB_MLX5DV_SET64(mkc, mkc, len, length);
+    UCT_IB_MLX5DV_SET(create_mkey_in, in, translations_octword_actual_size, list_size);
+
+    klm = UCT_IB_MLX5DV_ADDR_OF(create_mkey_in, in, klm_pas_mtt);
+    for(i = 0; i < list_size; i++) {
```
Comment (yosefe) [inline src/uct/ib/mlx5/ib_mlx5_dv.c]: space before (

File: src/uct/ib/mlx5/ib_mlx5_dv.c [hunk 10]
```diff
@@ -24,6 +33,110 @@ ucs_status_t uct_ib_mlx5dv_init_obj(uct_ib_mlx5dv_t *obj, uint64_t type)
...
+        if (i == list_size - 1) {
+            UCT_IB_MLX5DV_SET(klm, klm, byte_count, length % reg_length);
+        } else {
+            UCT_IB_MLX5DV_SET(klm, klm, byte_count, reg_length);
+        }
+        UCT_IB_MLX5DV_SET(klm, klm, mkey, mr->lkey);
+        UCT_IB_MLX5DV_SET64(klm, klm, address, addr + i * reg_length);
```
Comment (yosefe) [inline src/uct/ib/mlx5/ib_mlx5_dv.c]: ( ) around ` i * reg_length` expr

File: src/uct/ib/mlx5/ib_mlx5_dv.c [hunk 11]
```diff
@@ -24,6 +33,110 @@ ucs_status_t uct_ib_mlx5dv_init_obj(uct_ib_mlx5dv_t *obj, uint64_t type)
...
+    memh->atomic_dvmr = mlx5dv_devx_obj_create(md->dev.ibv_context, in, inlen,
+                                               out, sizeof(out));
+    if (memh->atomic_dvmr == NULL) {
+        ucs_error("CREATE_MKEY KSM failed: %m");
+        status = UCS_ERR_IO_ERROR;
+        goto out;
+    }
+
+    memh->super.atomic_rkey =
+        UCT_IB_MLX5DV_GET(create_mkey_out, out, mkey_index) << 8 |
```
Comment (yosefe) [inline src/uct/ib/mlx5/ib_mlx5_dv.c]: ( ) around << expr

File: src/uct/ib/base/ib_device.h [hunk 12]
```diff
@@ -61,7 +61,8 @@ enum {
     UCT_IB_DEVICE_FLAG_DC_V2    = UCS_BIT(7),   /* Device supports DC ver 2 */
     UCT_IB_DEVICE_FLAG_AV       = UCS_BIT(8),   /* Device supports compact AV */
     UCT_IB_DEVICE_FLAG_DC       = UCT_IB_DEVICE_FLAG_DC_V1 |
-                                  UCT_IB_DEVICE_FLAG_DC_V2 /* Device supports DC */
+                                  UCT_IB_DEVICE_FLAG_DC_V2, /* Device supports DC */
+    UCT_IB_DEVICE_FLAG_KSM      = UCS_BIT(9)    /* Device supports KSM */
```
Comment (yosefe) [inline src/uct/ib/base/ib_device.h]: maybe it's time to extend uct_ib_md_t to uct_ib_mlx5_md_t, and add this flag there?

File: src/uct/ib/base/ib_device.c [hunk 13]
```diff
@@ -242,6 +242,8 @@ static void uct_ib_async_event_handler(int fd, void *arg)
     ibv_ack_async_event(&event);
 }
 
+extern uct_ib_md_ops_t uct_ib_verbs_md_ops;
```
[the remark this answers is not in the document]
Comment (Artemy-Mellanox) [reply src/uct/ib/base/ib_device.c]: ib_device wasn't dependent on ib_md till now. May move uct_ib_verbs_md_open to ib_md.c?

File: src/uct/ib/base/ib_device.h (hunk 12 above)
Comment (Artemy-Mellanox) [reply src/uct/ib/base/ib_device.h]: Need to check for KSM whether it's mlx5 or not.
Look here  https://github.com/openucx/ucx/pull/3095/files#diff-734c9882bf0ceede1a096f651a5f2204R93

File: src/uct/ib/base/ib_device.c (hunk 13 above)
[the remark this answers is not in the document]
Comment (yosefe) [reply src/uct/ib/base/ib_device.c]: "May move uct_ib_verbs_md_open to ib_md.c?" - yes

File: src/uct/ib/base/ib_device.h (hunk 12 above)
Comment (yosefe) [reply src/uct/ib/base/ib_device.h]: you mean that would need to add mlx5 check to the gtest? this is not a problem.

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/8650/ for details (Mellanox internal link).

**Pull request merged**